query/src/find.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+///! This module defines some core types that support find expressions: sources,
+///! variables, expressions, etc.
+///! These are produced as 'fuel' by the query parser, consumed by the query
+///! translator and executor.
+///!
+///! Many of these types are defined as simple structs that are little more than
+///! a richer type alias: a variable, for example, is really just a fancy kind
+///! of string.
+///!
+///! At some point in the future, we might consider reducing copying and memory
+///! usage by recasting all of these string-holding structs and enums in terms
+///! of string references, with those references being slices of some parsed
+///! input query string, and valid for the lifetime of that string.
+///!
+///! For now, for the sake of simplicity, all of these strings are heap-allocated.
+///!
+///! Furthermore, we might cut out some of the chaff here: each time a 'tagged'
+///! type is used within an enum, we have an opportunity to simplify and use the
+///! inner type directly in conjunction with matching on the enum. Before diving
+///! deeply into this it's worth recognizing that this loss of 'sovereignty' is
+///! a tradeoff against well-typed function signatures and other such boundaries.
+
+
+// TODO: support other kinds of sources.
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub enum SrcVar {
+    DefaultSrc,
+}
+
+pub enum Constant {}    // This is essentially Box. TODO: flesh out.
+
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub struct Variable {
+    pub name: String,
+}
+
+pub enum FnArg {
+    Constant { constant: Constant },
+    Variable { variable: Variable },
+    Src      { src: SrcVar },
+}
+
+pub enum PullPattern {
+    Constant { constant: Constant },
+    Variable { variable: Variable },
+}
+
+pub struct Pull {
+    pub src: SrcVar,
+    pub var: Variable,
+    pub pattern: PullPattern,      // Constant, variable, or plain variable.
+}
+
+pub struct Aggregate {
+    pub fn_name: String,
+    pub args: Vec<FnArg>,
+}
+
+// TODO: look up the idiomatic way to express these kinds of type
+// combinations in Rust. It must be common in ASTs. Trait objects
+// presumably aren't the answer…
+pub enum Element {
+    Variable { variable: Variable },
```

Comment:
https://doc.rust-lang.org/1.1.0/syntax/ast/enum.Expr_.html

… looks like this is the idiomatic way to express ASTs!